src/main/kotlin/com/github/avrokotlin/avro4k/encoder/RecordEncoder.kt

```diff
@@ -1,110 +1,49 @@
-      }
-   }
-}
-
-@ExperimentalSerializationApi
-class RecordEncoder(private val schema: Schema,
-                    override val serializersModule: SerializersModule,
-                    val callback: (Record) -> Unit) : AbstractEncoder(), StructureEncoder {
-
-   private val builder = RecordBuilder(schema)
-   private var currentIndex = -1
-
-   override fun fieldSchema(): Schema {
-      // if the element is nullable, then we should have a union schema which we can extract the non-null schema from
-      val currentFieldSchema = schema.fields[currentIndex].schema()
-      return if (currentFieldSchema.isNullable) {
-         currentFieldSchema.extractNonNull()
-      } else {
-         currentFieldSchema
-      }
-   }
-
-   override fun addValue(value: Any) {
-      builder.add(value)
-   }
-
-   override fun encodeString(value: String) {
-      builder.add(StringToAvroValue.toValue(fieldSchema(), value))
-   }
-
-   override fun encodeValue(value: Any) {
-      builder.add(value)
-   }
-
-   override fun encodeElement(descriptor: SerialDescriptor, index: Int): Boolean {
-      currentIndex = index
-      return true
-   }
-
-   override fun encodeByteArray(buffer: ByteBuffer) {
-      builder.add(buffer)
-   }
-
-   override fun encodeFixed(fixed: GenericFixed) {
-      builder.add(fixed)
-   }
-
-   override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) {
-      builder.add(ValueToEnum.toValue(fieldSchema(), enumDescriptor, index))
-   }
-
-   override fun beginStructure(descriptor: SerialDescriptor): CompositeEncoder {
-      // if we have a value type, then we don't want to begin a new structure
-      return if (AnnotationExtractor(descriptor.annotations).valueType())
-         this
-      else
-         super<StructureEncoder>.beginStructure(descriptor)
-   }
-
-   override fun endStructure(descriptor: SerialDescriptor) {
-      callback(builder.record())
-   }
-
-   override fun encodeNull() {
-      builder.add(null)
-   }
+class RecordEncoder(
+        private val recordSchema: Schema,
+        override val serializersModule: SerializersModule,
+        override val configuration: AvroConfiguration,
+        private val callback: (GenericRecord) -> Unit,
+) : AvroStructureEncoder() {
+    private val builder: GenericRecordBuilder
+    private lateinit var currentField: Field
+
+    init {
+        recordSchema.ensureOfType(Schema.Type.RECORD)
+        builder = GenericRecordBuilder(recordSchema)
+    }
+
+    override val currentUnresolvedSchema: Schema get() = currentField.schema()
+
+    override fun encodeNativeValue(value: Any?) {
+        builder.set(currentField, value)
+    }
+
+    override fun shouldEncodeElement(descriptor: SerialDescriptor, index: Int): Boolean {
+        val fieldName = descriptor.getElementAvroName(configuration.namingStrategy, index)
+        val foundField = recordSchema.getField(fieldName)
```

Comment:
> fields of kotlin classes are not forced to be in same order of avro schema fields

Regarding performances, accessing an array from index (kotlin fields must be in same order as avro schema) or accessing a hashtable with a string hash, is it different ? I yes, we can detect good order during the beginStructure by returning a `WellSortedRecordEncoder` or `NotWellSortedRecordEncoder`